numpyro/infer/elbo.py

```diff
@@ -73,6 +80,104 @@ def single_particle_elbo(rng_key):
+                      "To eliminate this warning, ensure model and guide sites "
+                      "occur in the same order.\n" +
+                      "Model sites:\n  " + "\n  ".join(model_sites) +
+                      "Guide sites:\n  " + "\n  ".join(guide_sites))
+
+
+class TraceMeanField_ELBO(ELBO):
+    """
+    A trace implementation of ELBO-based SVI. This is currently the only
+    ELBO estimator in NumPyro that uses analytic KL divergences when those
+    are available.
+
+    .. warning:: This estimator may give incorrect results if the mean-field
+        condition is not satisfied.
+    """
+    def loss(self, rng_key, param_map, model, guide, *args, **kwargs):
+        """
+        Evaluates the ELBO with an estimator that uses num_particles many samples/particles.
+
+        :param jax.random.PRNGKey rng_key: random number generator seed.
+        :param dict param_map: dictionary of current parameter values keyed by site
+            name.
+        :param model: Python callable with NumPyro primitives for the model.
+        :param guide: Python callable with NumPyro primitives for the guide.
+        :param args: arguments to the model / guide (these can possibly vary during
+            the course of fitting).
+        :param kwargs: keyword arguments to the model / guide (these can possibly vary
+            during the course of fitting).
+        :return: negative of the Evidence Lower Bound (ELBO) to be minimized.
+        """
+        def single_particle_elbo(rng_key):
+            model_seed, guide_seed = random.split(rng_key)
+            seeded_model = seed(model, model_seed)
+            seeded_guide = seed(guide, guide_seed)
+            subs_guide = substitute(seeded_guide, data=param_map)
+            guide_trace = trace(subs_guide).get_trace(*args, **kwargs)
+            subs_model = substitute(replay(seeded_model, guide_trace), data=param_map)
+            model_trace = trace(subs_model).get_trace(*args, **kwargs)
+            _check_mean_field_requirement(model_trace, guide_trace)
+
+            elbo_particle = 0
+            for name, model_site in model_trace.items():
+                if model_site["type"] == "sample" and not isinstance(model_site["fn"], dist.PRNGIdentity):
+                    if model_site["is_observed"]:
+                        log_prob = jnp.sum(model_site["fn"].log_prob(model_site["value"]))
+                        log_prob = scale_and_mask(log_prob, scale=model_site["scale"])
+                        elbo_particle = elbo_particle + log_prob
+                    else:
+                        guide_site = guide_trace[name]
+                        try:
+                            kl_qp = kl_divergence(guide_site["fn"], model_site["fn"])
+                            kl_qp = scale_and_mask(kl_qp, scale=guide_site["scale"])
+                            if isinstance(kl_qp, (int, float)):
+                                kl_qp_sum = kl_qp * np.prod(guide_site["fn"].batch_shape)
```

Comment:
what's going on here? doesn't `kl_qp` include`guide_site["fn"].batch_shape` already?

Reply:
I thought so too. This follows [this line](https://github.com/pyro-ppl/pyro/blob/dev/pyro/infer/trace_mean_field_elbo.py#L108) in Pyro. I guess we need it for [KL[masked, masked]](https://github.com/pyro-ppl/numpyro/pull/748/files#diff-a54d4ede5b490415bf09bbe4c4dd8103R169) where it returns `0.` when `mask=False`. There might be other places which I don't have knowledge about. Nevertheless, I believe we can remove this logic by simply using `kl_qp_sum = jnp.sum(kl_qp)` here without if/else. WDYT?

Reply:
Anyway, I just remove this logic. It is better. Thanks!

Reply:
hmm. @fritzo i feel like we may have pair-coded that line... do you perhaps remember why...?

Reply:
There must have been a shortcut where we returned a broadcastable scalar value rather than a batched tensor, and the `np.prod` is required to get the power right when summing over many identical tensor elements. However I cannot think of a case other than `{-math.inf, 0, math.inf}`, and in those cases the power has no effect. You might simply convert this to an assertion like
```py
assert (not isinstance(kl_qp, (int, float)) or (kl_qp in (-math.inf, 0, math.inf))
```
(Aside: I wish Python had an "implies" operator so we didn't have to write "a implies b" as "(not a) or b")